Source/Diffusion.cpp

```diff
@@ -378,214 +399,254 @@ Diffusion::diffuse_scalar (const Vector<MultiFab*>&  S_old,
     opn.setDomainBC(mlmg_lobc, mlmg_hibc);
     opnp1.setDomainBC(mlmg_lobc, mlmg_hibc);
 
-    for (int icomp=0; icomp<num_comp; ++icomp)
+    if (verbose)
     {
-       if (verbose)
-       {
-          amrex::Print() << "diffusing scalar "<< icomp+1 << " of " << num_comp << "\n"
-                         << "rho flag "<< rho_flag << "\n";
-       }
+       amrex::Print() << "diffusing of " << nComp << " scalars \n"
+                      << "rho flag "<< rho_flag << "\n";
+    }
 
-       int sigma = S_comp + icomp;
+    int sigma = S_comp;
 
-       if (is_diffusive[icomp] == 0)
-       {
-          for (int n = 0; n < AMREX_SPACEDIM; n++)
-          {
-            if (fluxn[n]!=0 && fluxnp1[n]!=0)
-            {
-              fluxn[n]->setVal(0.0,fluxComp+icomp,1);
-              fluxnp1[n]->setVal(0.0,fluxComp+icomp,1);
-            }
-          }
-          break;
-       }
-
-       if (add_old_time_divFlux && be_cn_theta!=1)
-       {
-          Real a = 0.0;
-          Real b = -(1.0-be_cn_theta)*dt;
+    if (add_old_time_divFlux && be_cn_theta!=1) {
+        Real a = 0.0;
+        Real b = -(1.0-be_cn_theta)*dt;
 
-          if(verbose)
+        if(verbose)
              Print()<<"Adding old time diff ...\n";
 
-          {
-            if (has_coarse_data)
-            {
-              MultiFab::Copy(*Solnc,*S_old[1],sigma,0,1,0);
-              if (rho_flag == 2)
-              {
-                MultiFab::Divide(*Solnc,*Rho_old[1],Rho_comp,0,1,0);
-              }
-              opn.setCoarseFineBC(Solnc.get(), cratio[0]);
+        {
+            if (has_coarse_data) {
+                MultiFab::Copy(*Solnc,*S_old[1],sigma,0,nComp,0);
+                if (rho_flag == 2) {
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (Gpu::notInLaunchRegion())
+#endif
+                    for (MFIter mfi(*Solnc, TilingIfNotGPU()); mfi.isValid(); ++mfi)
+                    {
+                        const Box& bx = mfi.tilebox();
+                        auto const& sol_arr = Solnc->array(mfi);
+                        auto const& rho_arr = Rho_old[1]->const_array(mfi,Rho_comp);
+                        amrex::ParallelFor(bx, [sol_arr, rho_arr, nComp]
```

Comment:
@esclapez Why not use the 4D loop and /= to save a few lines of code?

Reply:
A habits of minimizing the number of divisions. Here I do only one instead of nComp, but if you prefer having more granularity on the ParallelFor, I can change that.

Reply:
Since PeleLM does way more components than IAMR, I'd imagine you'd see the biggest differences there (if there's any difference at all). I was just wondering.